server/util.go

```diff
@@ -500,6 +502,39 @@ func DefaultBaseappOptions(appOpts types.AppOptions) []func(*baseapp.BaseApp) {
 	}
 }
 
+// readChainIdFromHome reads chain id from home directory.
+func readChainIdFromHome(homeDir string) (string, error) {
+	cfg := tmcfg.DefaultConfig()
+	cfg.SetRoot(homeDir)
+
+	// if the node's current height is not zero then try to read the chainID from comet db.
+	db, err := node.DefaultDBProvider(&node.DBContext{ID: "blockstore", Config: cfg})
+	if err != nil {
+		return "", err
+	}
+
+	blockStore := tmstore.NewBlockStore(db)
+	defer func() {
+		if err := blockStore.Close(); err != nil {
+			panic(err)
+		}
+	}()
+
+	// if the blockStore.LoadBaseMeta() is nil (no blocks are created/synced so far), fallback to genesis chain-id
+	if blockStore.LoadBaseMeta() == nil {
+		appGenesis, err := tmtypes.GenesisDocFromFile(filepath.Join(homeDir, "config", "genesis.json"))
+		if err != nil {
+			return "", err
+		}
+
+		return appGenesis.ChainID, nil
+	}
+
+	chainID := blockStore.LoadBaseMeta().Header.ChainID
+
+	return chainID, nil
```

Comment:
I think it's better to call blockStore.LoadBaseMeta one time
```suggestion
	// if the blockStore.LoadBaseMeta() is nil (no blocks are created/synced so far), fallback to genesis chain-id
	baseMeta := blockStore.LoadBaseMeta()
	if baseMeta != nil {
		return baseMeta.Header.ChainID, nil
	}
	if appGenesis, err := tmtypes.GenesisDocFromFile(filepath.Join(homeDir, "config", "genesis.json")); err == nil {
		return appGenesis.ChainID, nil
	}
	return "", err
```